# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [2]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

os.getcwd()


'/Users/charleschan/python/springboard/u6'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
df = pd.read_csv(os.getcwd() + '/data/step3_output.csv')
df.head()

Unnamed: 0                           Name       state  summit_elev  \
0           2               Hilltop Ski Area      Alaska         2090   
1           4            Sunrise Park Resort     Arizona        11100   
2           5  Yosemite Ski & Snowboard Area  California         7800   
3          10               Donner Ski Ranch  California         8012   
4          12                  June Mountain  California        10090   

   vertical_drop  fastQuads  quad  triple  double  surface  ...  \
0            294          0     0       1       0        2  ...   
1           1800          1     2       3       1        0  ...   
2            600          0     0       1       3        1  ...   
3            750          0     0       1       5        2  ...   
4           2590          2     0       0       4        1  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0            30.0        150.000000       36.0   
1              800.0            80.0        115.000000       49.0   
2               88.0             0.0        110.000000       84.0   
3              505.0            60.0        163.000000       82.0   
4             1500.0             0.0        115.103943       58.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0     30.000000      34.00000              152.0   
1            250.0     74.000000      78.00000              104.0   
2            300.0     47.000000      47.00000              107.0   
3            400.0     75.000000      75.00000              170.0   
4            250.0     57.916957      64.16681              128.0   

   NightSkiing_ac  clusters  
0            30.0         0  
1            80.0         1  
2             0.0         1  
3             0.0         1  
4             0.0         1  

[5 rows x 24 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [4]:
#testx = df.select_dtypes(include=['object']) # select object type columns
#print(testx.info())


df = pd.concat([df.drop(columns=['state'], axis=1), pd.get_dummies(df['state'])], axis=1)


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 167 entries, 0 to 166
Data columns (total 58 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         167 non-null    int64  
 1   Name               167 non-null    object 
 2   summit_elev        167 non-null    int64  
 3   vertical_drop      167 non-null    int64  
 4   fastQuads          167 non-null    int64  
 5   quad               167 non-null    int64  
 6   triple             167 non-null    int64  
 7   double             167 non-null    int64  
 8   surface            167 non-null    int64  
 9   total_chairs       167 non-null    int64  
 10  Runs               167 non-null    float64
 11  TerrainParks       167 non-null    float64
 12  LongestRun_mi      167 non-null    float64
 13  SkiableTerrain_ac  167 non-null    float64
 14  Snow Making_ac     167 non-null    float64
 15  daysOpenLastYear   167 non-null    float64
 16  yearsOpen          167 non

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [6]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [7]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)


Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [8]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)



**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [9]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)


## Review Model Outcomes — Iterate over additional models as needed

In [10]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error

explained_variance_score(y_test, y_pred)




-2.6701566126395327e+22

In [11]:
mean_absolute_error(y_test, y_pred)



592846401190.492

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [12]:
print(lm.intercept_)




149099094965.11618


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [13]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


Coefficient
Unnamed: 0         1.988621e+00
summit_elev        2.242212e+00
vertical_drop      7.958531e-01
fastQuads          4.015370e+12
quad               5.049335e+12
triple             8.350415e+12
double             9.784724e+12
surface            1.011097e+13
total_chairs       1.769903e+13
Runs               1.291992e+00
TerrainParks       6.035156e-01
LongestRun_mi      7.842865e-01
SkiableTerrain_ac  3.890991e-01
Snow Making_ac     1.768066e+00
daysOpenLastYear   3.852539e-01
yearsOpen          5.590820e-01
averageSnowfall    3.515625e-02
AdultWeekday       9.713867e+00
projectedDaysOpen  3.059082e-01
NightSkiing_ac     8.297119e-01
clusters           1.093628e+00
Alaska             6.364129e+11
Arizona            6.364129e+11
California         1.761686e+12
Colorado           1.405813e+12
Connecticut        1.405813e+12
Idaho              1.535230e+12
Illinois           1.095639e+12
Indiana            8.973087e+11
Iowa               1.095639e+12
Maine              1.405813e+12
Maryland           1.180619e+10
Massachusetts      1.261272e+12
Michigan           2.358589e+12
Minnesota          1.761686e+12
Missouri           8.973087e+11
Montana            1.095639e+12
Nevada             6.364129e+11
New Hampshire      2.046179e+12
New Jersey         3.721721e+07
New Mexico         1.535230e+12
New York           2.619344e+12
North Carolina     1.095639e+12
Ohio               1.261272e+12
Oregon             1.095639e+12
Pennsylvania       1.957198e+12
Rhode Island       8.681471e+04
South Dakota       6.364129e+11
Tennessee          6.364129e+11
Utah               1.095639e+12
Vermont            1.653080e+12
Virginia           8.973087e+11
Washington         1.261272e+12
West Virginia      1.095639e+12
Wisconsin          1.957198e+12
Wyoming            1.405813e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [17]:
df_no_state = pd.read_csv(os.getcwd() + '/data/step3_output.csv')

In [19]:
df_no_state.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 167 entries, 0 to 166
Data columns (total 24 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         167 non-null    int64  
 1   Name               167 non-null    object 
 2   state              167 non-null    object 
 3   summit_elev        167 non-null    int64  
 4   vertical_drop      167 non-null    int64  
 5   fastQuads          167 non-null    int64  
 6   quad               167 non-null    int64  
 7   triple             167 non-null    int64  
 8   double             167 non-null    int64  
 9   surface            167 non-null    int64  
 10  total_chairs       167 non-null    int64  
 11  Runs               167 non-null    float64
 12  TerrainParks       167 non-null    float64
 13  LongestRun_mi      167 non-null    float64
 14  SkiableTerrain_ac  167 non-null    float64
 15  Snow Making_ac     167 non-null    float64
 16  daysOpenLastYear   167 non

In [20]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df_no_state.drop(['Name','AdultWeekend','state'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df_no_state['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [21]:

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)


In [22]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [23]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

In [24]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model


explained_variance_score(y_test, y_pred)


0.5258528582587472

In [25]:
mean_absolute_error(y_test, y_pred)


7.227882046530307

In [26]:
print(lm.intercept_)


56.188959607855125


In [27]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


Coefficient
Unnamed: 0            0.574867
summit_elev           2.993821
vertical_drop         1.628218
fastQuads             0.065074
quad                  0.372379
triple                1.119886
double                0.492185
surface               1.044833
total_chairs          0.295050
Runs                  0.945590
TerrainParks          0.122433
LongestRun_mi         0.431317
SkiableTerrain_ac     0.365445
Snow Making_ac        1.657960
daysOpenLastYear      1.879994
yearsOpen             0.071191
averageSnowfall       1.413166
AdultWeekday         10.662985
projectedDaysOpen     0.793697
NightSkiing_ac        0.479849
clusters              3.242070

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [31]:
df_no_state_elev = pd.read_csv(os.getcwd() + '/data/step3_output.csv')
df_no_state_elev.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 167 entries, 0 to 166
Data columns (total 24 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         167 non-null    int64  
 1   Name               167 non-null    object 
 2   state              167 non-null    object 
 3   summit_elev        167 non-null    int64  
 4   vertical_drop      167 non-null    int64  
 5   fastQuads          167 non-null    int64  
 6   quad               167 non-null    int64  
 7   triple             167 non-null    int64  
 8   double             167 non-null    int64  
 9   surface            167 non-null    int64  
 10  total_chairs       167 non-null    int64  
 11  Runs               167 non-null    float64
 12  TerrainParks       167 non-null    float64
 13  LongestRun_mi      167 non-null    float64
 14  SkiableTerrain_ac  167 non-null    float64
 15  Snow Making_ac     167 non-null    float64
 16  daysOpenLastYear   167 non

In [32]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df_no_state_elev.drop(['Name','AdultWeekend','state', 'summit_elev'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df_no_state_elev['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [33]:

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)


In [34]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [35]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

In [36]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model


explained_variance_score(y_test, y_pred)


0.4867575712917872

In [37]:
mean_absolute_error(y_test, y_pred)


7.771152242082643

In [38]:
print(lm.intercept_)


56.11978101747415


In [39]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


Coefficient
Unnamed: 0            0.522299
vertical_drop         2.143313
fastQuads             0.231019
quad                  0.290778
triple                0.994340
double                0.635271
surface               0.756773
total_chairs          0.418554
Runs                  0.124621
TerrainParks          0.409303
LongestRun_mi         0.332597
SkiableTerrain_ac     1.663626
Snow Making_ac        2.463512
daysOpenLastYear      1.428257
yearsOpen             0.241475
averageSnowfall       2.063462
AdultWeekday         10.293628
projectedDaysOpen     0.529357
NightSkiing_ac        0.459468
clusters              2.736264

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -2.6701566126395327e+22 | 592846401190.492 |-|
| Model 2. | 0.5258528582587472 | 7.227882046530307 |'state'|
| Model 3. | 0.4867575712917872 | 7.771152242082643 |'state','summit_elev','base_elev'|

Model Selection:

In [0]:
Model 3